### PR TITLE
Derive image tag from operator version

### LINKED ISSUE #11287: Determine Keycloak image tag from Operator version
Currently the operator by default uses `latest` tag for Keycloak images. To avoid hardcoding the tag in every release, the operator can derivate the default matching Keycloak image tag from the operator version.

Proposed solution:
* Remove [`operator.keycloak.image`](https://github.com/keycloak/keycloak/blob/1661a4ecc76fee3ff86d8d8f5d3c91728b333662/operator/src/main/resources/application.properties#L8) config option.
* Add a new config option: `operator.keycloak.image.name=quay.io/keycloak/keycloak`.
* Add a new config option: `operator.keycloak.image.tag` (this will not be defined by default in [`application.properties`](https://github.com/keycloak/keycloak/blob/1661a4ecc76fee3ff86d8d8f5d3c91728b333662/operator/src/main/resources/application.properties).
* Use the following flow when [determining the Keycloak image](https://github.com/keycloak/keycloak/blob/1661a4ecc76fee3ff86d8d8f5d3c91728b333662/operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java#L406):
   * If `image` field is defined in the CR, use that.
   * If `operator.keycloak.image.tag` is defined by the user, use `operator.keycloak.image.name:operator.keycloak.image.tag`.
   * Otherwise, use `operator.keycloak.image.name:{calculatedTag}`. Calculate the tag as follows:
```
if (quarkus.application.version == "999-SNAPSHOT") {
   calculatedTag = "nightly"
} else {
   calculatedTag = quarkus.application.version
}
```

### PR DESCRIPTION
**Pull request by andreaTP**
Resolves: #11287 

Manually tested since the important bit is when the application version changes due to the release in pom files.

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/Config.java [hunk 1]
```diff
@@ -27,7 +30,32 @@ public interface Config {
...
     interface Keycloak {
-        String image();
+        Image image();
         String imagePullPolicy();
+
+        interface Image {
+            String name();
+            Optional<String> tag();
+
+            default String getFinalTag() {
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/Config.java:40]: I wonder how to test this, i.e. how much can we trust `quarkus.application.version`.

But I think we should at least have a unit test for this. @andreaTP Can you please add one?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Thanks, @andreaTP.

Can you please also check the failed test? Other than that LGTM.

File: operator/src/main/resources/application.properties [hunk 2]
```diff
@@ -5,5 +5,5 @@ quarkus.container-image.builder=jib
 quarkus.operator-sdk.crd.validate=false
 
 # Operator config
-operator.keycloak.image=quay.io/keycloak/keycloak:latest
+operator.keycloak.image.name=quay.io/keycloak/keycloak
 operator.keycloak.image-pull-policy=Always
```
Comment (vmuzikar) [inline operator/src/main/resources/application.properties:9]: One more thing. I think we could also move the pull policy under image to keep it consistent, i.e. `operator.keycloak.image.pull-policy`. WDYT?

Comment (andreaTP) [reply operator/src/main/resources/application.properties:9]: Mixed feelings, I feel like `image-pull-policy` reflects best the property in Kubernetes (which is a separate field `imagePullPolicy`), but not really fussy about it.

**Comment by vmuzikar** [comment]
A summary of offline discussion with @andreaTP: we will piggyback on the OLM tests to actually test `quarkus.application.version` behaves correctly in our use-case. For that we need #11247 to be merged first.

**Comment by andreaTP** [comment]
Moving this PR back to draft due to ☝️

**Comment by andreaTP** [comment]
Superseded by #11511

**Pull request closed without merging**